controller/konnect/reconciler_konnectapiauth.go

```diff
@@ -39,8 +55,26 @@ func NewKonnectAPIAuthConfigurationReconciler(
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *KonnectAPIAuthConfigurationReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	secretLabelPredicate, err := predicate.LabelSelectorPredicate(
+		metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				SecretCredentialLabel: SecretCredentialLabelValueKonnect,
+			},
```

Comment:
What happens if such a label is removed for any reason? the `updatePredicate` is not triggered, as the new object does not have the label. We may circumvent this by explicitly setting the `updatePredicate` to check the old object has the label.

Also, I can't find the place where you add such a label to the referenced secret. Did I miss it?

Reply:
Then, the user in order to set a secret as the token source has to do both the actions:
- set the proper field in the `KonnectAPIAuthConfiguration` resource, and
- label the secret.

If one of the above is not performed, the secret is not taken into account as a token container.

Is that correct?

Reply:
Yes. This way we efficiently watch the `Secrets` (via the label selector).

Returning an error will enforce this so that users have the best possible experience and we don't support the use case (no label) with suboptimal UX. WDYT?

Reply:
I think that requiring these two actions to the user is not the best UX and can be confusing. We could attach such a label from the controller logic, instead of asking the user to do it.

In the first reconciliation loop, we fetch the secret and attach a label, then we ensure such a label is always there. We have to be careful though, as the problem I raised in the first comment of this thread has to be avoided.

WDYT?

Reply:
I see your point but I believe that this will only complicate the code and make the operator more difficult to be used in gitops setups (this label will need to be ignored as it will not be present in user applied manifests).

We already do use this label in KIC (this change was introduced in 3.0 https://docs.konghq.com/kubernetes-ingress-controller/latest/guides/migrate/credential-kongcredtype-label/ to improve the performance and make the credential secrets explicitly marked as such) and I believe KGO should use the same approach.

Reply:
> I see your point but I believe that this will only complicate the code and make the operator more difficult to be used in gitops setups (this label will need to be ignored as it will not be present in user applied manifests).

Ok, it makes sense.

I still find this approach a bit odd though, and I wonder: do we really need such an optimization? I understand there can be a massive amount of secrets in the cluster, but they do not likely change often. I can see the point of such an improvement for highly mutable objects that may have frequent status updates (pods for example), but secrets do not have status at all and they should be pretty stable objects. Can you help me understand why we really need it?

Reply:
This has been raised quite some time ago by one of KIC users in https://github.com/Kong/kubernetes-ingress-controller/issues/2431.

Do note that this was about webhook validation which can only use label and not fields (as it was done prior to solving that issue).

What we're discussing here is filtering at the controller-runtime predicates level but if we ever need to implement any additional logic elsewhere (say in the aforementioned webhook) we'll hit a similar issue: how to efficiently list if we don't have the label to select on.

> I understand there can be a massive amount of secrets in the cluster, but they do not likely change often. I can see the point of such an improvement for highly mutable objects that may have frequent status updates (pods for example), but secrets do not have status at all and they should be pretty stable objects.

That might be an assumption which is not always true or more importantly: it's not really relevant when we e.g. list `Secret`s that can be used as credentials.

I'm happy to discuss this on a Zoom / Slack 👍

Reply:
> Do note that this was about webhook validation which can only use label and not fields (as it was done prior to solving that issue).

I see the root problem, and how it was solved, but I really think this is not a good UX. Having a webhook on secrets is a very bad idea at first instance: if the secret is malformed for a consumer we should reflect this in the consumer status in my opinion. But I know that we really LOVE webhooks and we want to have them all over our code.

But this is a different use-case: we are doing it just for optimization purpose, and as I said above I don't see how this is problematic. 

> That might be an assumption which is not always true

Can you give me an example of frequently changing secrets? I can't find one 🤔 

> we e.g. list Secrets that can be used as credentials.

We have the secret reference in the `KonnectAPIAuthConfiguration`, we don't need to list them.

Reply:
> if the secret is malformed for a consumer we should reflect this in the consumer status in my opinion

We already had this discussion w.r.t to a different feature that had rejecting at webhook level invalid configuration. This came straight from our users who want to prevent invalid configuration to get into the cluster.

We can always evangelize best practices internally and externally ( so that e.g. users use objects status for monitoring instead of trying to block invalid configuration at the webhook level). It doesn't seem that we do that (enough), since users are calling for the latter. Do we have an alternative for them? Do we understand their use case and have a solution for that?

Having said that: ideally that would be configurable when people don't want this level of validation. Here, IMHO configurability would only make this simple feature, a complicated one.

> But I know that we really LOVE webhooks and we want to have them all over our code.

That's not a feeling shared by all involved 🙃 

> Can you give me an example of frequently changing secrets? I can't find one 🤔

It doesn't really how to be the use case of frequently changing secrets. Let's try to make a bit less contrived example: a user or an internal PM asks for Konnect token rotation operator. There's 1 requirement: all Konnect tokens embedded either in `KonnectAPIAuthConfiguration` or in `Secret`s has to be checked 7 days ahead of expiration and rotated. Token in `Secret`s should be rotated **regardless** if they are being used or not.

For this use case we'd need to list all `Secret`s in the cluster. Having the label we can easily filter only those that we care about.

---

Additionally, without the label, `listKonnectAPIAuthConfigurationsReferencingSecret` is being called for each `Secret`. Each create or update on a `Secret` in a cluster will cause 

- a list operation of `KonnectAPIAuthConfiguration`s
- check if any of the items in the returned listm, references the `Secret`

Reply:
> My impression is that we are speculating on something that could happen in the future, and providing a solution before having an actual request or issue.

This way we save ourselves from a breaking change in the future.

I believe this is the right direction and this approach has a valuable trait: we can  relax the requirement without causing a breaking change.